File: src/new-apps/android-tracker/app/src/test/java/com/geovault/tracker/data/TrackerGeometryMergePolicyTest.kt
package com.geovault.tracker.data

import com.geovault.tracker.GeoJsonLineString
import com.geovault.tracker.Tracker
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Test

class TrackerGeometryMergePolicyTest {

    @Test
    fun merged_whenNoExisting_returnsIncoming() {
        val incoming = tracker(
            id = "t1",
            name = "Incoming",
            geometryCoords = listOf(listOf(1.0, 2.0)),
            color = "#FF0000",
        )

        val merged = TrackerGeometryMergePolicy.merged(existing = null, incoming = incoming)

        assertEquals("t1", merged.id)
        assertEquals("Incoming", merged.name)
        assertEquals("#FF0000", merged.color)
        assertEquals(1, merged.geometry?.coordinates?.size)
    }

    @Test
    fun merged_preservesExistingMetadata_whenIncomingOmitsFields() {
        val existing = tracker(
            id = "t1",
            name = "Existing Name",
            geometryCoords = listOf(listOf(10.0, 20.0)),
            color = "#112233",
            ownerEmail = "owner@example.com",
            updatedAt = 100L,
        )
        val incoming = tracker(
            id = "t1",
            name = "",
            geometryCoords = listOf(listOf(30.0, 40.0)),
            color = null,
            ownerEmail = null,
            updatedAt = null,
        )

        val merged = TrackerGeometryMergePolicy.merged(existing = existing, incoming = incoming)

        assertEquals("Existing Name", merged.name)
        assertEquals("#112233", merged.color)
        assertEquals("owner@example.com", merged.owner_email)
        assertEquals(100L, merged.updated_at)
        assertNotNull(merged.geometry)
        assertEquals(listOf(listOf(30.0, 40.0)), merged.geometry?.coordinates)
    }

    @Test
    fun merged_usesIncomingValues_whenProvided() {
        val existing = tracker(
            id = "t1",
            name = "Old",
            geometryCoords = listOf(listOf(1.0, 1.0)),
            color = "#000000",
            ownerEmail = "old@example.com",
            updatedAt = 10L,
        )
        val incoming = tracker(
            id = "t1",
            name = "New",
            geometryCoords = listOf(listOf(2.0, 2.0)),
            color = "#ABCDEF",
            ownerEmail = "new@example.com",
            updatedAt = 20L,
        )

        val merged = TrackerGeometryMergePolicy.merged(existing = existing, incoming = incoming)

        assertEquals("New", merged.name)
        assertEquals("#ABCDEF", merged.color)
        assertEquals("new@example.com", merged.owner_email)
        assertEquals(20L, merged.updated_at)
        assertEquals(listOf(listOf(2.0, 2.0)), merged.geometry?.coordinates)
    }

    private fun tracker(
        id: String,
        name: String,
        geometryCoords: List<List<Double>>?,
        color: String?,
        ownerEmail: String? = null,
        updatedAt: Long? = null,
    ): Tracker {
        return Tracker(
            id = id,
            name = name,
            color = color,
            settings = null,
            geometry = geometryCoords?.let { GeoJsonLineString(type = "LineString", coordinates = it) },
            point_params = null,
            last_point = null,
            bbox = null,
            tracker_secret = null,
            created_at = null,
            subscribed_at = null,
            updated_at = updatedAt,
            is_owner = null,
            visibility = null,
            share_params_with_recipients = null,
            share_params_with_world = null,
            owner_email = ownerEmail,
            subscriber_count = null,
            world_share_id = null,
            world_share_url = null,
            shared_with_emails = null
        )
    }
}

File: src/backend/extensions/live_track/src/frontend/test/trackerGeometryMergePolicy.test.mjs
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { mergeTrackerGeometry } from '../src/trackerGeometryMergePolicy.ts';

test('merge preserves existing metadata when incoming omits fields', () => {
  const existing = {
    id: 't1',
    name: 'Existing',
    color: '#112233',
    geometry: { type: 'LineString', coordinates: [[1, 2, 10]] },
    point_params: [{ acc: 1 }],
    owner_email: 'owner@example.com',
  };
  const incoming = {
    id: 't1',
    name: '',
    geometry: { type: 'LineString', coordinates: [[3, 4, 20]] },
  };
  const merged = mergeTrackerGeometry(existing, incoming);
  assert.equal(merged.name, 'Existing');
  assert.equal(merged.color, '#112233');
  assert.equal(merged.owner_email, 'owner@example.com');
  assert.deepEqual(merged.geometry.coordinates, [[3, 4, 20]]);
  assert.deepEqual(merged.point_params, [{ acc: 1 }]);
});

test('merge keeps longer existing point_params when incoming list is shorter without geometry', () => {
  const existing = {
    id: 't1',
    name: 'Existing',
    geometry: { type: 'LineString', coordinates: [[1, 2]] },
    point_params: [{ acc: 1 }, { acc: 2 }],
  };
  const incoming = {
    id: 't1',
    name: 'Existing',
    point_params: [{ acc: 9 }],
  };
  const merged = mergeTrackerGeometry(existing, incoming);
  assert.equal(merged.point_params.length, 2);
});
